### PR TITLE
Add clang builds: Ubuntu, CentOS 7/8 (+ ossl), macOS, Debian9 and Windows-msys2 (from #1684)

### LINKED ISSUE #1129: ci: Build with clang-cl
## Description

I couldn't find any open issue for this and since the Thunderbird guys are building with clang-cl, we should probably add this to CI.

(I'm planning to tackle this myself, but if anyone is already working on this let me know!)

### LINKED ISSUE #1668: clang/libgcrypt v1.9.3 issue  (CentOS 8 workflow)
**Target:**  CentOS 8
**Tests**   (normal, 2.3.1, clang, clang++)
**Module:**  libgcrypt / 1.9.3
**Error:**
```
basic.c:315:16: error: inline assembly requires more registers than available
  asm volatile("movdqu %[data0], %%xmm0\n"
               ^
basic.c:315:16: error: inline assembly requires more registers than available
```

Other tests use  libgcrypt /v1.8.7  and  libgcrypt / v1.9.0-beta277  which are compiled without issues.

_Originally posted by @maxirmx in https://github.com/rnpgp/rnp/issues/1129#issuecomment-959510564_

### LINKED ISSUE #1669: clang/GnuPG v2.3.0-beta983 issue (CentOS 8 workflow)
**Target:** CentOS 8
**Tests** (normal, beta, clang, clang++),  (sanitize, beta, clang, clang++)
**Module:** GnuPG v2.3.0-beta983
**Error:**
```
libcommon.a(libcommon_a-iobuf.o):(.bss+0x8): multiple definition of `iobuf_debug_mode'
t-iobuf.o:(.bss+0x0): first defined here
``` 
Other tests use GnuPG v2.2.24 whicj compiles without issues.

_Originally posted by @maxirmx in https://github.com/rnpgp/rnp/issues/1129#issuecomment-959545493_

### LINKED ISSUE #1670: clang/rnp build issue (CentOS 7 workflow)
Target: CentOS 7
Tests pkgconfig-cmake-target (no, clang, clang++), pkgconfig-cmake-target (yes, clang, clang++
Module: librepgp

```
In file included from /__w/rnp/rnp/src/librepgp/stream-common.cpp:50:
In file included from /__w/rnp/rnp/src/lib/crypto/mem.h:34:
In file included from /usr/include/botan-2/botan/secmem.h:12:
/usr/include/botan-2/botan/mem_ops.h:147:18: error: no member named 'is_trivially_copyable' in namespace 'std'
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(T), "");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/botan-2/botan/mem_ops.h:124:45: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
                                       ~~~~~^
/usr/include/botan-2/botan/mem_ops.h:147:46: error: 'T' does not refer to a value
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(T), "");
                                             ^
/usr/include/botan-2/botan/mem_ops.h:124:67: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
                                                                  ^
/usr/include/botan-2/botan/mem_ops.h:145:19: note: declared here
template<typename T> inline void typecast_copy(uint8_t out[], T in[], size_t N)
                  ^
/usr/include/botan-2/botan/mem_ops.h:147:18: error: no member named 'value' in the global namespace
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(T), "");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/botan-2/botan/mem_ops.h:124:71: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
                                                                    ~~^
/usr/include/botan-2/botan/mem_ops.h:170:18: error: no member named 'is_trivially_copyable' in namespace 'std'
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(From) && std::is_trivial<To>::value, "");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/botan-2/botan/mem_ops.h:124:45: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
                                       ~~~~~^
/usr/include/botan-2/botan/mem_ops.h:170:46: error: 'From' does not refer to a value
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(From) && std::is_trivial<To>::value, "");
                                             ^
/usr/include/botan-2/botan/mem_ops.h:124:67: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
                                                                  ^
/usr/include/botan-2/botan/mem_ops.h:168:27: note: declared here
template <class To, class From> inline To typecast_copy(const From *src) noexcept
                          ^
/usr/include/botan-2/botan/mem_ops.h:170:18: error: no member named 'value' in the global namespace
   static_assert(BOTAN_IS_TRIVIALLY_COPYABLE(From) && std::is_trivial<To>::value, "");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/botan-2/botan/mem_ops.h:124:71: note: expanded from macro 'BOTAN_IS_TRIVIALLY_COPYABLE'
#define BOTAN_IS_TRIVIALLY_COPYABLE(T) std::is_trivially_copyable<T>::value
```

_Originally posted by @maxirmx in https://github.com/rnpgp/rnp/issues/1129#issuecomment-959670336_

### PR DESCRIPTION
**Pull request by ronaldtse**
Originally from #1684 by @maxirmx .

Clang build and test for the following workflows:

- centos7
- centos8-ossl
- centos8
- debian9
- macos
- ubuntu
- windows (msys) 

Fixed:

- Fixes #1129
- Fixes #1670
- Fixes #1669
- Fixes #1668

Changes previously reviewed but not included in this PR:

- #1685 
- #1686

### CODE REVIEW

**Comment by ronaldtse** [comment]
CentOS 8 build failed due to this:

```
+ curl -L -o cmake.sh https://github.com/Kitware/CMake/releases/download/v3.20.5/cmake-3.20.5-Linux-x86_64.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   661  100   661    0     0  22033      0 --:--:-- --:--:-- --:--:-- 22033

  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:04 --:--:--     0
100   212  100   212    0     0     41      0  0:00:05  0:00:05 --:--:--    74
100   212  100   212    0     0     41      0  0:00:05  0:00:05 --:--:--    74
+ PREFIX=/usr
+ mkdir -p /usr
+ sudo sh cmake.sh --skip-license --prefix=/usr
cmake.sh: line 1: syntax error near unexpected token `<'
cmake.sh: line 1: `﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>ServerBusy</Code><Message>Egress is over the account limit.'
Error: Process completed with exit code 2.
```

The fetch for CMake from GitHub went over the traffic limit. The `curl` script needs to handle retries... but might be tricky: https://stackoverflow.com/questions/24512366/retry-timeout-curl

**Comment by ronaldtse** [comment]
"centos7 / centos7 tests stable": Additional failure here that we have previously seen about LeakSanitizer.

```
 33/239 Test  #26: rnp_tests.test_cli_rnp_keyfile ................................................***Failed   35.08 sec
Note: Google Test filter = rnp_tests.test_cli_rnp_keyfile
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from rnp_tests
[ RUN      ] rnp_tests.test_cli_rnp_keyfile
Good signature made Wed Nov 10 09:03:30 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [SC]
      090bd712a1166be572252c3c9747d2a6b3a63124
uid           key-merge-uid-2
uid           key-merge-uid-1
Signature(s) verified successfully
Good signature made Wed Nov 10 09:03:39 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [SC]
      090bd712a1166be572252c3c9747d2a6b3a63124
uid           key-merge-uid-2
uid           key-merge-uid-1
Signature(s) verified successfully
[signature_check() /__w/rnp/rnp/src/librepgp/stream-sig.cpp:227] invalid or untrusted key
BAD signature made Wed Nov 10 09:03:39 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [ESCA]
      090bd712a1166be572252c3c9747d2a6b3a63124
Signature verification failure: 1 invalid signature(s), 0 unknown signature(s)
Hello world!
This is a sample file to use for signing.
Operation failed: No suitable key
[init_encrypted_src() /__w/rnp/rnp/src/librepgp/stream-parse.cpp:2149] failed to obtain decrypting key or password
Hello world!
This is a sample file to use for signing.
[       OK ] rnp_tests.test_cli_rnp_keyfile (35009 ms)
[----------] 1 test from rnp_tests (35009 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (35010 ms total)
[  PASSED  ] 1 test.
==69866==LeakSanitizer has encountered a fatal error.
==69866==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==69866==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

**Comment by ronaldtse** [comment]
The last GHA re-run failed at "centos7 / centos7 tests stable" with LeakSanitizer at the same test:

```
 33/239 Test  #26: rnp_tests.test_cli_rnp_keyfile ................................................***Failed   35.08 sec
Note: Google Test filter = rnp_tests.test_cli_rnp_keyfile
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from rnp_tests
[ RUN      ] rnp_tests.test_cli_rnp_keyfile
Good signature made Wed Nov 10 09:03:30 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [SC]
      090bd712a1166be572252c3c9747d2a6b3a63124
uid           key-merge-uid-2
uid           key-merge-uid-1
Signature(s) verified successfully
Good signature made Wed Nov 10 09:03:39 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [SC]
      090bd712a1166be572252c3c9747d2a6b3a63124
uid           key-merge-uid-2
uid           key-merge-uid-1
Signature(s) verified successfully
[signature_check() /__w/rnp/rnp/src/librepgp/stream-sig.cpp:227] invalid or untrusted key
BAD signature made Wed Nov 10 09:03:39 2021
using RSA (Encrypt or Sign) key 9747d2a6b3a63124

pub   3072/RSA (Encrypt or Sign) 9747d2a6b3a63124 2018-08-31 [ESCA]
      090bd712a1166be572252c3c9747d2a6b3a63124
Signature verification failure: 1 invalid signature(s), 0 unknown signature(s)
Hello world!
This is a sample file to use for signing.
Operation failed: No suitable key
[init_encrypted_src() /__w/rnp/rnp/src/librepgp/stream-parse.cpp:2149] failed to obtain decrypting key or password
Hello world!
This is a sample file to use for signing.
[       OK ] rnp_tests.test_cli_rnp_keyfile (35009 ms)
[----------] 1 test from rnp_tests (35009 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (35010 ms total)
[  PASSED  ] 1 test.
==69866==LeakSanitizer has encountered a fatal error.
==69866==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==69866==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

File: .github/workflows/debian9.yml [hunk 1]
```diff
@@ -37,31 +40,37 @@ jobs:
...
-      
+
+# I386: actions/checkout@V2,  actions/cache@V2 actions/cache@V2
+#       give error about OCI container [https://github.com/actions/checkout/issues/334]
+#                                      [https://github.com/actions/runner/issues/1011]
       - name: Check out repository
         uses: actions/checkout@v1
         with:
-          fetch-depth: 1
+          fetch-depth: 10
```
Comment (ribose-jeffreylau) [inline .github/workflows/debian9.yml:65]: Is there a reason to fetch 10 commits?

File: ci/lib/install_functions.inc.sh [hunk 2]
```diff
@@ -522,6 +523,16 @@ linux_install_debian() {
     "${build_dependencies_deb[@]}" \
     "$@"
 
+  if [ "${CC-gcc}" = "clang" ]; then
+# Add apt.llvm.org repository and install clang
+# We may use https://packages.debian.org/stretch/clang-3.8 as well but this package gets installed to
+# /usr/lib/clang... and requires update-alternatives which would be very ugly considering CC/CXX environment
+# settings coming from yaml already
```
Comment (ribose-jeffreylau) [inline ci/lib/install_functions.inc.sh:530]: There are trailing spaces.

File: ci/lib/install_functions.inc.sh [hunk 3]
```diff
@@ -522,6 +523,16 @@ linux_install_debian() {
...
     "$@"
 
+  if [ "${CC-gcc}" = "clang" ]; then
+# Add apt.llvm.org repository and install clang
+# We may use https://packages.debian.org/stretch/clang-3.8 as well but this package gets installed to
+# /usr/lib/clang... and requires update-alternatives which would be very ugly considering CC/CXX environment
+# settings coming from yaml already   
+    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key|sudo apt-key add -
+    ${SUDO} apt-add-repository "deb http://apt.llvm.org/stretch/ llvm-toolchain-stretch main"
+    ${SUDO} apt-get install -y clang
```
Comment (ribose-jeffreylau) [inline ci/lib/install_functions.inc.sh:533]: There is a trailing space here.

File: ci/lib/install_functions.inc.sh [hunk 4]
```diff
@@ -546,19 +555,34 @@ msys_install() {
...
+    )
+  else
+   packages+=(clang64/mingw-w64-clang-x86_64-clang 
+              clang64/mingw-w64-clang-x86_64-openmp
+              clang64/mingw-w64-clang-x86_64-libc++
+              clang64/mingw-w64-clang-x86_64-libbotan
+              clang64/mingw-w64-clang-x86_64-libssp
+              clang64/mingw-w64-clang-x86_64-json-c
+              clang64/mingw-w64-clang-x86_64-libsystre
+   )
```
Comment (ribose-jeffreylau) [inline ci/lib/install_functions.inc.sh:573]: These lines need to indent by one more space.

File: ci/lib/install_functions.inc.sh [hunk 5]
```diff
@@ -546,19 +555,34 @@ msys_install() {
     make
     pkg-config
     mingw64/mingw-w64-x86_64-cmake
-    mingw64/mingw-w64-x86_64-gcc
-    mingw64/mingw-w64-x86_64-json-c
     mingw64/mingw-w64-x86_64-python3
   )
 
+  if [ "${CC}" = "gcc" ]; then
+    packages+=(mingw64/mingw-w64-x86_64-gcc
```
Comment (ribose-jeffreylau) [inline ci/lib/install_functions.inc.sh:562]: There is a trailing space.

File: ci/lib/install_functions.inc.sh [hunk 6]
```diff
@@ -546,19 +555,34 @@ msys_install() {
...
-    mingw64/mingw-w64-x86_64-json-c
     mingw64/mingw-w64-x86_64-python3
   )
 
+  if [ "${CC}" = "gcc" ]; then
+    packages+=(mingw64/mingw-w64-x86_64-gcc 
+               mingw64/mingw-w64-x86_64-json-c
+    )
+  else
+   packages+=(clang64/mingw-w64-clang-x86_64-clang
```
Comment (ribose-jeffreylau) [inline ci/lib/install_functions.inc.sh:566]: Also, a trailing space here.

**Review by ribose-jeffreylau — CHANGES_REQUESTED** [review]
Just some minor things.

File: .github/workflows/debian9.yml (hunk 1 above)
Comment (ronaldtse) [reply .github/workflows/debian9.yml:65]: I can't find the comment but @maxirmx commented before that on debian:9, fetch-depth of 1 will cause actions/checkout@v1 to fail. A fetch-depth of 10 worked.

**Comment by ronaldtse** [comment]
All tests are passing except for debian9 (which always run for nearly 2h).

File: ci/lib/install_functions.inc.sh (hunk 5 above)
Comment (ronaldtse) [reply ci/lib/install_functions.inc.sh]: Fixed.

File: ci/lib/install_functions.inc.sh (hunk 6 above)
Comment (ronaldtse) [reply ci/lib/install_functions.inc.sh]: Fixed.

File: ci/lib/install_functions.inc.sh (hunk 4 above)
Comment (ronaldtse) [reply ci/lib/install_functions.inc.sh]: Fixed.

File: ci/lib/install_functions.inc.sh (hunk 3 above)
Comment (ronaldtse) [reply ci/lib/install_functions.inc.sh]: Fixed.

File: ci/lib/install_functions.inc.sh (hunk 2 above)
Comment (ronaldtse) [reply ci/lib/install_functions.inc.sh]: Fixed.

**Comment by ronaldtse** [comment]
> The last GHA re-run failed at "centos7 / centos7 tests stable" with LeakSanitizer at the same test:

This is no longer failing. Apparently it is a transient failure.

**Comment by ronaldtse** [comment]
@ribose-jeffreylau if it's okay with you, once it passes feel free to approve and merge. Thanks!

**Pull request merged**